test/verify/check-realms

```diff
@@ -307,6 +307,96 @@ class CommonTests:
         # sometimes polling for info and joining a domain creates this noise
         self.allow_journal_messages('.*org.freedesktop.DBus.Error.Spawn.ChildExited.*')
 
+    def checkClientCertAuthentication(self):
+        '''Common tests for certificate authentication
+
+        This assumes that IdM and sssd are all set up correctly already.
+        '''
+        m = self.machine
+
+        # join domain, wait until it works
+        m.execute("printf '[cockpit.lan]\\nfully-qualified-names = no\\n' >> /etc/realmd.conf")
+        m.execute("echo %s | realm join -vU %s cockpit.lan" % (self.admin_password, self.admin_user), timeout=300)
+        m.execute('while ! id alice; do sleep 5; systemctl restart sssd; done', timeout=300)
+
+        # upload certificates to client machine
+        m.upload(["alice.pem", "alice.key", "bob.pem", "bob.key"], "/var/tmp",
+                 relative_dir="src/tls/ca/")
+        alice_cert_key = ['--cert', "/var/tmp/alice.pem", '--key', "/var/tmp/alice.key"]
+        alice_user_pass = ['-u', 'alice:' + self.alice_password]
+        bob_cert_key = ['--cert', "/var/tmp/bob.pem", '--key', "/var/tmp/bob.key"]
+
+        # ensure sssd certificate lookup works
+        user_obj = m.execute('busctl call org.freedesktop.sssd.infopipe /org/freedesktop/sssd/infopipe/Users '
+                             'org.freedesktop.sssd.infopipe.Users FindByCertificate s -- '
+                             '''"$(cat /var/tmp/alice.pem)" | sed 's/^o "//; s/"$//' ''')
+        self.assertEqual(m.execute('busctl get-property org.freedesktop.sssd.infopipe ' + user_obj.strip() +
+                                   ' org.freedesktop.sssd.infopipe.Users.User name').strip(),
+                         's "alice"')
+
+        # These tests have to be run with curl, as chromium-headless does not support selecting/handling client-side
+        # certificates; it just rejects cert requests. For interactive tests, grab src/tls/ca/alice.p12 and import
+        # it into the browser.
+
+        def do_test(authopts, expected, not_expected=[], session_leader=None):
+            m.start_cockpit(tls=True)
+            output = m.execute(['curl', '-ksS', '-D-'] + authopts + ['https://localhost:9090/cockpit/login'])
+            for s in expected:
+                self.assertIn(s, output)
+            for s in not_expected:
+                self.assertNotIn(s, output)
+
+            if session_leader:
+                tries = 0
+                while tries < 10:
+                    out = m.execute('loginctl show-user --property=Sessions alice')
+                    sessions = out.split('=')[1].split()
+                    if len(sessions) == 1:
+                        break
+                    time.sleep(5)
```

Comment:
I remember that back then when I wrote this (many moons ago now..), starting with a single `m.execute("until ...; do sleep")` loop, it wasn't working. Primarily because it simultaneously needs to wait for the number and the status of sessions, and a simple

    m.execute('until [ "$(loginctl show-user --property=State --value alice)" = "active" ]; do sleep 1; done')

wasn't enough. But in check-login I then came up with that "iterate over all sessions and find the active one" approach, and that seems to work robustly at least locally for me.

So I used the shell `until; sleep; done` loop in both tests now, and also synced the two tests. Let's see what CI says about the stability of this (it runs each check-login/realms test three times now, and they must all succeed).